src/gg.regression.unity.bots/Runtime/Scripts/Validation/RGValidationScript.cs

```diff
@@ -0,0 +1,251 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using StateRecorder.BotSegments.Models.SegmentValidations;
+using UnityEngine;
+
+namespace RegressionGames.Validation
+{
+    public class RGValidationScript
+    {
+
+        private bool _isPaused;
+        
+        public class ValidatorData
+        {
+            public MethodInfo Method { get; set; }
+            public int Frequency { get; set; }
+            
+            public ValidationMode Mode { get; set; }
+            
+            public SegmentValidationStatus Status { get; set; } = SegmentValidationStatus.UNKNOWN;
+            
+            public System.Exception ThrownException { get; set; }
+        }
+
+        public List<ValidatorData> Validators = new ();
+        
+        /**
+         * Called when any validation here changes
+         */
+        public delegate void ValidationsUpdated();
+        public event ValidationsUpdated OnValidationsUpdated;
+
+        private string _className;
+        private int frame = 0;
+        private ValidatorData currentValidator;
+        private SegmentValidationStatus currentStatus = SegmentValidationStatus.UNKNOWN;
+
+        public void Initialize()
+        {
+
+            _className = GetType().FullName;
+            
+            // Cache all methods with UpdateMethod attribute
+            Validators = new List<ValidatorData>();
+            var methods = GetType().GetMethods(BindingFlags.Instance | 
+                                               BindingFlags.NonPublic | 
+                                               BindingFlags.Public);
+        
+            foreach (var method in methods)
+            {
+                var attribute = method.GetCustomAttribute<RGValidate>();
+                if (attribute != null)
+                {
+                    Validators.Add(new ValidatorData 
+                    { 
+                        Method = method,
+                        Frequency = attribute.Frequency,
+                        Mode = attribute.Mode
+                    });
+                    RGDebug.Log("[RGValidator] Activated " + method.Name);
+                } 
+            }
+            
+        }
+
+        public void AssertAsTrue(string message = null)
```

Comment:
It took me a minute to understand how currentStatus can be updated within ProcessValidations. Please confirm that the idea is for the validators to call these functions? If so would you mind documenting that?

Reply:
Yes that is correct! Added some more docs.